In [1]:
class ExpenseSharing:
     def __init__(self, friends):
         self.friends = friends
         self.expenses = {friend: 0 for friend in friends}

     def add_expense(self, payer, amount, participants):
         split_amount = amount / len(participants)
         for participant in participants:
             if participant != payer:
                 self.expenses[participant] += split_amount
                 self.expenses[payer] -= split_amount

     def calculate_settlement(self):  
         for friend, balance in self.expenses.items():
             if balance > 0:
                print(f"{friend} needs to be reimbursed: Rs.{balance:.2f}")
             elif balance < 0:
                print(f"{friend} owes: Rs.{-balance:.2f}")
             else:
                print(f"{friend} is settled up.")

In [2]:
if __name__ == "__main__":
    friends = ["Vasu", "Kavya", "Vino", "Siva", "Zara"]
    expense_sharing = ExpenseSharing(friends)

#Add some sample expenses:
    expense_sharing.add_expense("Vasu", 500, ["Vasu", "Kavya", "Vino", "Siva", "Zara"]) #paid for accommadation
    expense_sharing.add_expense("Kavya", 250, ["Vasu", "Kavya", "Vino"]) #paid for meals
    expense_sharing.add_expense("Vino", 150, ["Kavya" , "Vino", "Siva"]) #paid for fuel
    expense_sharing.add_expense("Siva", 300, ["Vasu", "Siva"])
    expense_sharing.add_expense("Zara", 200, ["Vasu", "Kavya", "Zara"])
    expense_sharing.add_expense("Vasu", 100, ["Vasu", "Zara"])
    expense_sharing.add_expense("Kavya", 400, ["Kavya", "Vino", "Zara"])
    
#Calculate settlement:
    print("\nFinal Settlement:")
    expense_sharing.calculate_settlement()


Final Settlement:
Vasu owes: Rs.150.00
Kavya owes: Rs.216.67
Vino needs to be reimbursed: Rs.216.67
Siva is settled up.
Zara needs to be reimbursed: Rs.150.00


Who owes Whom

In [3]:
class ExpenseSharing:
     def __init__(self, friends):
         self.friends = friends
         self.balances= {friend: 0 for friend in friends}

     def add_expense(self, payer, amount, participants):
         split_amount = amount / len(participants)
         for participant in participants:
             self.balances[participant] -= split_amount
         self.balances[payer] += amount

     def calculate_settlement(self):
         creditors = []
         debtors = []

         for friend, balance in self.balances.items():
             if balance > 0:
                 creditors.append((friend, balance))
             elif balance < 0:
                 debtors.append((friend, -balance))

         while debtors and creditors:
             debtor, debt_amount = debtors.pop()
             creditor, credit_amount = creditors.pop()

             payment = min(debt_amount, credit_amount)

             print(f"{debtor} owes {creditor}: Rs. {payment:.2f}")

             if debt_amount > payment:
                 debtors.append((debtor, debt_amount - payment))
             if credit_amount > payment:
                 creditors.append((creditor, credit_amount - payment))     

In [4]:
if __name__ == "__main__":
    friends = ["Vasu", "Kavya", "Vino", "Siva", "Zara"]
    expense_sharing = ExpenseSharing(friends)

# Add some sample expenses:
    expense_sharing.add_expense("Vasu", 500, ["Vasu", "Kavya", "Vino", "Siva", "Zara"]) #paid for accommadation
    expense_sharing.add_expense("Kavya", 250, ["Vasu", "Kavya", "Vino"]) #paid for meals
    expense_sharing.add_expense("Vino", 150, ["Kavya" , "Vino", "Siva"]) #paid for fuel
    expense_sharing.add_expense("Siva", 300, ["Vasu", "Siva"])
    expense_sharing.add_expense("Zara", 200, ["Vasu", "Kavya", "Zara"])
    expense_sharing.add_expense("Vasu", 100, ["Vasu", "Zara"])
    expense_sharing.add_expense("Kavya", 400, ["Kavya", "Vino", "Zara"])
    
# Calculate settlement:
    print("\nFinal Settlement:")
    expense_sharing.calculate_settlement()


Final Settlement:
Zara owes Kavya: Rs. 150.00
Vino owes Kavya: Rs. 66.67
Vino owes Vasu: Rs. 150.00


In [ ]:
if __name__ == "__main__":

    friends = input("Enter the names of friends, seperated by commas:").split(",")
    friends = [friend.strip() for friend in friends]

    expense_sharing = ExpenseSharing(friends)

    while True:
        payer = input("Enter the name of the person who paid(of 'done' to finish):")
        if payer.lower() == "done":
            break

        amount = float(input("Enter the amount paid:")) 

        participants = input("Enter the names of participants for this expense, seperated by commas:").split(",")
        participants = [participants.strip() for participant in participants]

        expense_sharing.add_expense(payer, amount, participants)

    print("\nFinal Settlement")
    expense_sharing.calculate_settlement()

Calculations for every expenses

Vasu paid 500,
participants = 5,
500/5 = 100each,
Vasu paid 500, so vasu paid 400 more than his own share.

Kavya paid 250,
participants = 3,
250/3 = 83.33each,
Kavya paid 250, but her own share is only 83.33. So, she paid approximately 166.67 extra.


Vino paid 150, participants = 3, 150/3 = 50each, Vino paid 150, but his own share is 50. So, Vino paid 100extra.

Siva paid 300, participants = 2, 300/2 = 150each, Siva paid 300, and his own share is 150. So, Siva paid 150extra.

Zara paid 200, participants = 3, 200/3 = 66.67each, Zara paid 150, but her own share is 66.67. So, Zara paid 133.33extra.

Vasu paid 100, participants = 2, 100/2 = 50each, Vasu paid 100, but his own share is 50. So, Vasu paid 50extra.

Kavya paid 400, participants = 3, 400/3 = 133.33each, Kavya paid 150, but her own share is 133.33. So, Kavya paid 266.67extra.

Now combines Everything:

we calculate:

Net balance = Total amount paid - Total personal share.

Vasu paid 500:

500 + 100 = 600

Vasu's total share = 100+83.33+150+66.67+50 = 450

600-450 = 150 Vasu should receive = 150

Kavya paid 250:

250+ 400 = 650

Kavya's total share = 100+83.33+50+66.67+133.33 = 433.33

650-433.33 = 216.67 Kavya should receive = 216.67



Vino paid 150:

Vino's total share = 100+83.33+50+133.33 = 366.67

150-366.67 = -216.67 The minus sign means Vino didn't pay enough. Vino needs to pay 216.67



Siva paid 300:

Siva's total share = 100+150+50 = 300

300-300 = 0 Siva is settled up.



Zara paid 200:

Zara's total share = 100+50+66.67+133.33 = 350

200-350 = -150 Zara needs to pay 150



The Final Insight is:

People who should RECEIVE money:

Vasu 150

Kavya 216.67

People who should PAY money:

Vino 216.67

Zara 150

Already Settled:

Siva 0

Money to receive: 

150+216.67 = 366.67

Money to pay:

216.67+150 = 366.67

Everything balances